### PR TITLE
Disallow curried dependent context function types

### PR DESCRIPTION
**Pull request by odersky**
We can't implement them yet. Better to disallow them outright than to try map them
to  something wrong and have obscure follow-on errors.

Also: Implement returns in methods with context function result types

Previously, a return in a method with a CFT result type crashed the typer.

### CODE REVIEW

**Comment by odersky** [comment]
@nicolasstucki If you produce a new bootstrap compiler it would be good to have this PR in it as well. That would allow me to continue with my experiment to use CFTs in the compiler. Right now that's blocked since some methods that would have a CFT result type also have return statements.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1437,17 +1445,40 @@ class Typer extends Namer
...
+                  i"return mismatch from $owner, tparams = $tparams, locals = ${locals.toList}%, %")
+            info.instantiate(tparams.map(_.typeRef)).finalResultType
+          case info =>
+            info.finalResultType
+        def iftParamss = ctx.owner.ownersIterator
+          .filter(_.is(Method, butNot = Accessor))
+          .takeWhile(_.isAnonymousFunction)
+          .toList
+          .reverse
+          .map(_.paramSymss.head)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1479]: This map can be done before the `toList`

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:1479]: Yes, but that's not necessarily better, or is it? We are talking of iterators of length 1 - 2 on average

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:1479]: Then the extra allocation should be fine.

**Comment by nicolasstucki** [comment]
This change will be available in the compiler once #9311 is merged.

**Pull request merged**